### PR TITLE
Fixes issue #388 (nested dictionaries causing segfault)

### PR DESCRIPTION
**Pull request by Northern-Lights**
`yr_object_copy` was missing a case in the switch which should have initialized the `prototype_item` of the dictionary object.  This was causing a segfault for nested dictionaries when it came time for them to be copied.

I also added an `assert(FALSE)` as the default case, because if we don't know what we're copying, we should fail.

### CODE REVIEW

File: libyara/object.c [hunk 1]
```diff
@@ -609,6 +609,28 @@ int yr_object_copy(
       break;
 
     case OBJECT_TYPE_ARRAY:
+      ;
+      YR_OBJECT_ARRAY *array = (YR_OBJECT_ARRAY *) object;
```
Comment (plusvic) [inline libyara/object.c]: Declaring variable inside switch statements doesn't work in C89. You can use the "o" variable.

File: libyara/object.c [hunk 2]
```diff
@@ -609,6 +609,28 @@ int yr_object_copy(
...
+      YR_OBJECT *array_prototype_item;
+
+      yr_object_copy(array->prototype_item, &array_prototype_item);
+
+      ((YR_OBJECT_ARRAY *)copy)->prototype_item = array_prototype_item;
+
+      if (array->items == NULL)
```
Comment (plusvic) [inline libyara/object.c]: You don't need to initialize array->items here. This should be enough:

```
      yr_object_copy(((YR_OBJECT_ARRAY*) object)->prototype_item, &o);
      ((YR_OBJECT_ARRAY*) copy)->prototype_item = o
```

File: libyara/object.c [hunk 3]
```diff
@@ -628,6 +650,21 @@ int yr_object_copy(
       }
 
       break;
+
+    case OBJECT_TYPE_DICTIONARY:
+      ;
+      YR_OBJECT_DICTIONARY *dict = (YR_OBJECT_DICTIONARY *) object;
```
Comment (plusvic) [inline libyara/object.c]: Same thing here, invalid variable declaration in C89. This should work:

```
      yr_object_copy(((YR_OBJECT_DICTIONARY*) object)->prototype_item, &o);
      ((YR_OBJECT_DICTIONARY*) copy)->prototype_item = o;
```

**Comment by plusvic** [comment]
Very good job!

**Comment by Northern-Lights** [comment]
Changes implemented.  Thanks for the suggestions!

**Pull request merged**